src/main/java/org/cubedb/core/MultiCubeImpl.java

```diff
@@ -46,7 +46,9 @@ private Cube createNewCube(String cubeName) {
 
 	@Override
 	public void insert(List<DataRow> data) {
-		Lock l = lock.writeLock();
+		Lock l = lock.readLock(); // Although we are technically writing here,
```

Comment:
After a longer discussion we found out that:

a) there's no need to use complicated ReadWriteLock

b) the problem is in system stats, which catch the monitor lock